core/drivers/versal_nvm.c

```diff
@@ -0,0 +1,944 @@
+		return ret;
+
+	ret = versal_mbox_notify(&cmd, NULL, NULL);
+	if (ret)
+		EMSG("Mailbox error");
+
+	return ret;
+}
+
+static TEE_Result versal_nvm_read(struct versal_nvm_read_req *req)
+{
+	uint32_t *arg = NULL;
+	uint32_t val = 0;
+
+	if (!req)
+		return TEE_ERROR_GENERIC;
+
+	switch (req->efuse_id) {
+	case EFUSE_READ_DNA:
+	case EFUSE_READ_DEC_EFUSE_ONLY:
+	case EFUSE_READ_PUF_SEC_CTRL:
+	case EFUSE_READ_BOOT_ENV_CTRL:
+	case EFUSE_READ_SEC_CTRL:
+	case EFUSE_READ_MISC_CTRL:
+	case EFUSE_READ_SEC_MISC1:
+	case EFUSE_READ_USER_FUSES:
+	case EFUSE_READ_PUF_USER_FUSES:
+	case EFUSE_READ_PUF:
+		break;
+	case EFUSE_READ_OFFCHIP_REVOCATION_ID:
+		val = req->offchip_id;
+		arg = &val;
+		break;
+	case EFUSE_READ_REVOCATION_ID:
+		val = req->revocation_id;
+		arg = &val;
+		break;
+	case EFUSE_READ_IV:
+		val = req->iv_type;
+		arg = &val;
+		break;
+	case EFUSE_READ_PPK_HASH:
+		val = req->ppk_type;
+		arg = &val;
+		break;
+	case BBRAM_READ_USER_DATA:
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+	}
+
+	return efuse_req(req->efuse_id, req->ibuf, arg);
+}
+
+static TEE_Result versal_nvm_write(struct versal_nvm_write_req *req)
+{
+	uint32_t *arg = NULL;
+	uint32_t val = 0;
+
+	switch (req->efuse_id) {
+	case BBRAM_WRITE_AES_KEY:
+		val = req->bbram.aes_key_len;
+		arg = &val;
+		break;
+	case BBRAM_WRITE_USER_DATA:
+		val = req->bbram.user_data;
+		arg = &val;
+		break;
+	case EFUSE_PUF_USER_FUSE_WRITE:
+	case EFUSE_WRITE_PUF:
+	case EFUSE_WRITE:
+		break;
+	default:
+		return TEE_ERROR_GENERIC;
+	}
+
+	return efuse_req(req->efuse_id, req->ibuf, arg);
+}
+
+TEE_Result versal_efuse_read_dna(uint32_t *buf, size_t len)
+{
+	uint8_t lbuf[1024] __aligned_efuse = { 0 };
+	struct versal_nvm_read_req req = {
+		.efuse_id = EFUSE_READ_DNA,
+	};
+
+	if (len < EFUSE_DNA_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	req.ibuf[0].buf = lbuf;
+	req.ibuf[0].len = sizeof(lbuf);
+
+	if (versal_nvm_read(&req))
+		return TEE_ERROR_GENERIC;
+
+	memcpy(buf, lbuf, EFUSE_DNA_LEN);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_efuse_read_user_data(uint32_t *buf, size_t len,
+				       uint32_t first, size_t num)
+{
+	uint8_t lbuf[1024] __aligned_efuse = { 0 };
+	struct versal_efuse_user_data cfg __aligned_efuse = {
+		.addr = (uintptr_t)lbuf,
+		.start = first,
+		.num = num, /* fuses needs to be at least 40 bytes */
+	};
+	struct versal_nvm_read_req req = {
+		.efuse_id = EFUSE_READ_USER_FUSES,
+	};
+
+	if (first + num > EFUSE_MAX_USER_FUSES || len < num * sizeof(uint32_t))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	req.ibuf[0].buf = &cfg;
+	req.ibuf[0].len = sizeof(cfg);
+	req.ibuf[1].buf = lbuf;
+	req.ibuf[1].len = sizeof(lbuf);
+
+	cfg.addr = (paddr_t)virt_to_phys((void *)cfg.addr);
+
+	if (versal_nvm_read(&req))
+		return TEE_ERROR_GENERIC;
+
+	memcpy(buf, lbuf, num * sizeof(uint32_t));
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_efuse_read_iv(uint32_t *buf, size_t len,
+				enum versal_nvm_iv_type type)
+{
+	uint8_t lbuf[1024] __aligned_efuse = { 0 };
```

Comment:
That's quite a bit to store on the stack. Can it be done differently with a smaller buffer or by having the caller supply a well-aligned buffer? Another option is to allocate using the mempool API. The mempool API doesn't have a `mempool_memalign()`, but that shouldn't be hard to add.

There are quite a few other functions also doing large stack allocations. I'm not sure how big this problem is. If these are only called during boot initialization I guess we're safe, but if it can happen later during syscall processing then it's more concerning.

Reply:
I have the same comment on stack size. Stacks are few kByte by default. 1kB is quite a lot.